lib/compat/wordpress-6.1/persisted-preferences.php

```diff
@@ -0,0 +1,106 @@
+<?php
+/**
+ * Server-side requirements for database persisted preferences.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Register the user meta for persisted preferences.
+ */
+function gutenberg_register_persisted_preferences_meta() {
+	// Create a meta key that incorporates the blog prefix so that each site
+	// on a multisite can have distinct user preferences.
+	global $wpdb;
+	$meta_key = $wpdb->get_blog_prefix() . 'persisted_preferences';
+
+	register_meta(
+		'user',
+		$meta_key,
+		array(
+			'type'         => 'object',
+			'single'       => true,
+			'show_in_rest' => array(
+				'name'    => 'persisted_preferences',
+				'type'    => 'object',
+				'context' => array( 'edit' ),
+				'schema'  => array(
+					'type'                 => 'object',
+					'properties'           => array(
+						'_modified' => array(
+							'description' => __( 'The date and time the preferences were updated.', 'default' ),
+							'type'        => 'string',
+							'format'      => 'date-time',
+							'context'     => array( 'edit' ),
+							'readonly'    => true,
+						),
+					),
+					'additionalProperties' => true,
+				),
+			),
+		)
+	);
+}
+
+add_action( 'init', 'gutenberg_register_persisted_preferences_meta' );
+
+/**
+ * Configures the preferences package to use user meta persistence.
+ */
+function gutenberg_configure_persisted_preferences() {
+	$user_id = get_current_user_id();
+	if ( empty( $user_id ) ) {
+		return;
+	}
+
+	global $wpdb;
+	$meta_key = $wpdb->get_blog_prefix() . 'persisted_preferences';
+
+	$preload_data = get_user_meta( $user_id, $meta_key, true );
+
+	wp_add_inline_script(
+		'wp-preferences',
+		sprintf(
+			'( function() {
+				var serverData = %s;
+				var userId = "%s";
```

Comment:
Minor and belated, but it seems safer to encode `userId` formally, in the same way as `serverData`, even if we know that the result

```php
sprintf( 'userId = %s', wp_json_encode( $user_id ) );
```

would be identifical to the current form

```php
sprintf( 'userId = "%s"', $user_id );
```

The former doesn't rely on our own mental type checking, and redundancy helps with safety — especially around code interpolation. :)